modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/Diagram.js

```diff
@@ -26,38 +26,104 @@ const Diagram = ({
 	imageURL,
 	namespace,
 	navigationController,
+	newPinSettings,
+	pins,
 	spritemap,
 	zoomController,
 }) => {
+	const [addPinHandler, setAddPinHandler] = useState(false);
+	const [removePinHandler, setRemovePinHandler] = useState({
+		handler: false,
+		pin: null,
+	});
 	const [resetZoom, setResetZoom] = useState(false);
 	const [zoomInHandler, setZoomInHandler] = useState(false);
 	const [zoomOutHandler, setZoomOutHandler] = useState(false);
 	const [changedScale, setChangedScale] = useState(false);
 	const [scale, setScale] = useState(1);
 	const [selectedOption, setSelectedOption] = useState(1);
+	const [cPins, setCpins] = useState(pins);
+	const [showTooltip, setShowTooltip] = useState({
+		details: {
+			cx: 0,
+			cy: 0,
+			id: null,
+			label: '',
+			linked_to_sku: 'sku',
+			quantity: null,
+			sku: '',
+		},
+		tooltip: null,
+	});
+	const [addNewPinState, setAddNewPinState] = useState({
+		fill: newPinSettings.colorPicker.selectedColor,
+		radius: newPinSettings.defaultRadius,
+	});
+
+	useEffect(() => {
+		if (!showTooltip.tooltip) {
+			const newCPinState = cPins.map((element) => {
+				if (element.id === showTooltip.details.id) {
+					return {
+						cx: cPins[element.id].cx,
+						cy: cPins[element.id].cy,
+						draggable: cPins[element.id].draggable,
+						fill: cPins[element.id].fill,
+						id: showTooltip.details.id,
+						label: showTooltip.details.label,
+						linked_to_sku: showTooltip.details.linked_to_sku,
+						quantity: showTooltip.details.quantity,
+						r: cPins[element.id].r,
+						sku: showTooltip.details.sku,
+					};
+				}
+				else {
+					return element;
+				}
+			});
+			setCpins(newCPinState);
+		}
+		// eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
What is the reason for disabling this rule? I think we should be sure to include dependencies on any state he use in this effect

Reply:
It is generating a lot of problems updating every time cPins in this way. React doesn't work fine with d3 ever.

Reply:
> It is generating a lot of problems updating every time cPins in this way

This might be a bit of an indicator that the effect or state management in this component is wired up wrong. I can't tell without getting into the code too deeply, but often when we have to explicitly remove a dependency it means we are using that state in a non-reacty (I just made this word up) way.

Reply:
> React doesn't work fine with d3 ever

I'd also caution you on this. Yes, d3 is tricky to use, but it is definitely usable within React. I've always found that it ends up being how I have wired up my components or d3 logic that is the issue and not the relationship between react + d3. There are plenty of examples of people doing this well. Here are a few docs I have found helpful

https://blog.griddynamics.com/using-d3-js-with-react-js-an-8-step-comprehensive-manual/
Particularly this bit:
 > Step 2: Determine responsibilities


https://www.pluralsight.com/guides/using-d3.js-inside-a-react-app
Particularly how they talk about creating hooks to use d3, it helps break the code into chunks and see how it interacts with React.

[more on using d3 as custom hooks](https://compatt.medium.com/a-simple-dashboard-using-react-hooks-and-d3-1eca02ea0d18)